presto-prometheus/src/main/java/io/prestosql/plugin/prometheus/PrometheusClient.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.plugin.prometheus;
+
+import com.google.common.base.Supplier;
+import com.google.common.base.Suppliers;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.io.Resources;
+import io.airlift.json.JsonCodec;
+import io.prestosql.spi.type.DoubleType;
+import io.prestosql.spi.type.TimestampType;
+import io.prestosql.spi.type.TypeManager;
+import io.prestosql.spi.type.TypeSignature;
+import org.apache.http.client.utils.URIBuilder;
+
+import javax.inject.Inject;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.Objects.requireNonNull;
+
+public class PrometheusClient
+{
+    // schema name is fixed to "prometheus"
+    private final Supplier<Map<String, Object>> tableSupplier;
+    protected final PrometheusConfig config;
+    private static TypeManager typeManager;
+    protected static final String METRICS_ENDPOINT = "/api/v1/label/__name__/values";
+
+    @Inject
+    public PrometheusClient(PrometheusConfig config, JsonCodec<Map<String, Object>> metricCodec, TypeManager typeManager)
+            throws URISyntaxException
+    {
+        requireNonNull(config, "config is null");
+        requireNonNull(metricCodec, "metricCodec is null");
+        requireNonNull(typeManager, "typeManager is null");
+
+        tableSupplier = Suppliers.memoizeWithExpiration(metricsSupplier(metricCodec, getPrometheusMetricsURI(config)), 30, TimeUnit.SECONDS);
+        this.config = config;
+        this.typeManager = typeManager;
+    }
+
+    public Set<String> getSchemaNames()
+    {
+        return ImmutableSet.of("prometheus");
+    }
+
+    private URI getPrometheusMetricsURI(PrometheusConfig config)
+            throws URISyntaxException
+    {
+        // endpoint to retrieve metric names from Prometheus
+        URI uri = config.getPrometheusURI();
+        return new URIBuilder()
+                .setScheme(uri.getScheme())
+                .setHost(uri.getAuthority())
+                .setPath(uri.getPath().concat(METRICS_ENDPOINT))
```

Comment:
do we need the `uri.getPath()` in here? Can't we just `.setPath(METRICS_ENDPOINT)`?

Reply:
Currently, we're taking advantage of this to facilitate testing. Maybe it's not the best way to do it, but currently it allows to set the resource in test.

Reply:
I don't see how that changes this. Can you elaborate?

Reply:
We're creating a bogus URL in some tests and then special casing on the metrics endpoint in the override of `doGet` in the testing http server, PrometheusHttpServer.

So for instance in a test like `TestPrometheusClient#testHandleErrorResponse`, we `setPrometheusURI` to `prometheusHttpServer.resolve("/prometheus-data/prometheus-metrics-error.json"). The concat creates an odd URL (during test only) that looks like: `/prometheus-data/prometheus-metrics-error.json/api/v1/label/__name__/values` and later in `doGet` we split out the resource file the firing test actually wanted to use.

Reply:
Hmm, ok. I'll think on this.